### PR TITLE
Cache hardware thread count for <execution>

### LINKED ISSUE #1134: <execution>: cache thread::hardware_concurrency
Currently <execution> does not cache `hardware_concurrency` value, like it used to in certain configurations.

`GetNativeSystemInfo` may be a system call, so it may defeat the speedup obtained by parallel algorithms.

Consider caching it again. 

Note that since number of CPUs _may_ change at runtime, it may be a good idea to refresh the value after some `GetTickCount64` interval.

### PR DESCRIPTION
**Pull request by AlexGuteniev**
resolves #1134 (conservatively)
revert to what there were before atomic wait

### CODE REVIEW

File: stl/src/parallel_algorithms.cpp [hunk 1]
```diff
@@ -25,7 +25,13 @@ extern "C" {
 
 // TRANSITION, ABI
```
Comment (CaseyCarter) [inline stl/src/parallel_algorithms.cpp:26]: We can no longer remove this function in vNext without breaking the library, so this `TRANSITION` comment is no longer appropriate.

Comment (CaseyCarter) [reply stl/src/parallel_algorithms.cpp:26]:
```suggestion
```

**Comment by BillyONeal** [comment]
It looks like we can remove `#include <thread>` from `<execution>` now.

File: stl/src/parallel_algorithms.cpp [hunk 2]
```diff
@@ -23,9 +23,14 @@ namespace {
 
 extern "C" {
 
-// TRANSITION, ABI
 _NODISCARD unsigned int __stdcall __std_parallel_algorithms_hw_threads() noexcept {
-    return _STD thread::hardware_concurrency();
+    static int _Cached_hw_concurrency = -1;
+    unsigned int _Hw_concurrency      = __iso_volatile_load32(&_Cached_hw_concurrency);
```
Comment (BillyONeal) [inline stl/src/parallel_algorithms.cpp:28]:
```suggestion
    int _Hw_concurrency      = __iso_volatile_load32(&_Cached_hw_concurrency);

```

File: stl/src/parallel_algorithms.cpp [hunk 3]
```diff
@@ -23,9 +23,14 @@ namespace {
...
-// TRANSITION, ABI
 _NODISCARD unsigned int __stdcall __std_parallel_algorithms_hw_threads() noexcept {
-    return _STD thread::hardware_concurrency();
+    static int _Cached_hw_concurrency = -1;
+    unsigned int _Hw_concurrency      = __iso_volatile_load32(&_Cached_hw_concurrency);
+    if (_Hw_concurrency == -1) {
+        _Hw_concurrency = _STD thread::hardware_concurrency();
+        __iso_volatile_store32(&_Cached_hw_concurrency, _Hw_concurrency);
+    }
+    return _Hw_concurrency;
```
Comment (BillyONeal) [inline stl/src/parallel_algorithms.cpp:33]:
```suggestion
    static int _Cached_hw_concurrency = -1;
    int _Hw_concurrency      = __iso_volatile_load32(&_Cached_hw_concurrency);
    if (_Hw_concurrency == -1) {
        _Hw_concurrency = static_cast<int>(_STD thread::hardware_concurrency());
        __iso_volatile_store32(&_Cached_hw_concurrency, _Hw_concurrency);
    }

    return static_cast<unsigned int>(_Hw_concurrency);

```

**Comment by StephanTLavavej** [comment]
Thanks for restoring this performance optimization! :rocket:

**Pull request merged**